### PR TITLE
statsMgr information get for studio

### PR DESCRIPTION
**Pull request by xjlgod**
At present, Nebula Studio is doing multi task import. It need to know the total number of imported lines and the number of imported lines of each task.

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Comment (veezhang) [inline .idea/.gitignore:1]: please remove files in `.idea` directory.

File: pkg/stats/statsmgr.go [hunk 2]
```diff
@@ -11,19 +15,41 @@ import (
...
+	var wg sync.WaitGroup
+	wg.Add(numReadingFiles)
+	for _, file := range files {
+		path := file.FailDataPath
+		go func(path string) {
+			totalLine += countLines(path)
+			defer wg.Done()
+		}(*path)
+	}
+	wg.Wait()
```
Comment (veezhang) [inline pkg/stats/statsmgr.go:47]: There seems to be no need for multiple goroutine here?

File: pkg/stats/statsmgr.go [hunk 3]
```diff
@@ -92,3 +133,16 @@ func (s *StatsMgr) startWorker(numReadingFiles int) {
 		}
 	}
 }
+
+func countLines(path string) int64 {
```
Comment (veezhang) [inline pkg/stats/statsmgr.go:137]: This will cause the file to be read twice, will it affect performance?

File: pkg/stats/statsmgr.go [hunk 4]
```diff
@@ -92,3 +133,16 @@ func (s *StatsMgr) startWorker(numReadingFiles int) {
...
+	file, err := os.Open(path)
+	if err != nil {
+		return 0
```
Comment (veezhang) [inline pkg/stats/statsmgr.go:140]: Don't eat `err`?

File: pkg/stats/statsmgr.go [hunk 5]
```diff
@@ -11,19 +15,41 @@ import (
...
 	NumReadFailed int64
+	totalLine     int64
 	totalCount    int64
 	totalBatches  int64
 	totalLatency  int64
 	totalReqTime  int64
 }
 
-func NewStatsMgr(numReadingFiles int) *StatsMgr {
+type StatsQuery struct {
```
Comment (veezhang) [inline pkg/stats/statsmgr.go:28]: How about define `StatsMgr` as follows?
```go
type StatsMgr struct {
	StatsCh       chan base.Stats
	DoneCh        chan bool
	statsRW       sync.RWMutex
  stats         Stats
}

type Stats struct {
	NumFailed     int64
	NumReadFailed int64
	TotalLine     int64
	TotalCount    int64
	TotalBatches  int64
	TotalLatency  int64
	TotalReqTime  int64
}
```

Then, you can define a function `func (s *StatsMgr) Stats() Stats` replace the current `StatsQuery` function.

File: pkg/stats/statsmgr.go (hunk 3 above)
Comment (veezhang) [reply pkg/stats/statsmgr.go:137]: Will the size of the file be considered, not the number of lines in the file? Add `TotalSize`, `ProcessedSize`?

File: pkg/stats/statsmgr.go [hunk 6]
```diff
@@ -92,3 +133,17 @@ func (s *StatsMgr) startWorker(numReadingFiles int) {
...
+	file, err := os.Open(path)
+	if err != nil {
+		logger.Errorf("count line fail: %s", path)
+		return 0
+	}
+	fileScanner := bufio.NewScanner(file)
+	lineCount := int64(0)
+	for fileScanner.Scan() {
+		lineCount++
+	}
```
Comment (jievince) [inline pkg/stats/statsmgr.go:147]: Does it is faster? https://stackoverflow.com/a/24563853

**Review by jievince — COMMENTED** [review]
Good job!

File: pkg/stats/statsmgr.go [hunk 7]
```diff
@@ -92,3 +139,27 @@ func (s *StatsMgr) startWorker(numReadingFiles int) {
...
+func countLines(path string) (int64, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		logger.Errorf("count line fail: %s", path)
+		return 0, err
+	}
+	r := bufio.NewReader(file)
+	buf := make([]byte, 32*1024)
+	// There is no \n in the last line
+	count := 1
```
Comment (veezhang) [inline pkg/stats/statsmgr.go:152]: How do you know?
Maybe ?
```
a
b
```
or 
```
a
b

```

File: pkg/stats/statsmgr.go [hunk 8]
```diff
@@ -92,3 +139,27 @@ func (s *StatsMgr) startWorker(numReadingFiles int) {
...
+	lineSep := []byte{'\n'}
+	for {
+		c, err := r.Read(buf)
```
Comment (veezhang) [inline pkg/stats/statsmgr.go:155]: Is this normal when the file is larger than 3M?

File: pkg/stats/statsmgr.go [hunk 9]
```diff
@@ -92,3 +139,27 @@ func (s *StatsMgr) startWorker(numReadingFiles int) {
 		}
 	}
 }
+
+func countLines(path string) (int64, error) {
```
Comment (veezhang) [inline pkg/stats/statsmgr.go:143]: Should these be considered?
https://github.com/vesoft-inc/nebula-importer/blob/master/pkg/config/config.go#L87-L89

Comment (veezhang) [reply pkg/stats/statsmgr.go:143]: @jievince How do you think?

**Review by veezhang — APPROVED** [review]
LGTM

File: pkg/stats/statsmgr.go [hunk 10]
```diff
@@ -2,32 +2,62 @@ package stats
...
+func NewStatsMgr(files []*config.File) *StatsMgr {
+	numReadingFiles := len(files)
+	totalLine := int64(0)
+	var wg sync.WaitGroup
+	wg.Add(numReadingFiles)
+	for _, file := range files {
+		path := file.Path
+		withHeader := file.CSV.WithHeader
+		go func(path string) {
+			lines, err := csv.CountLines(path, *withHeader)
```
Comment (yixinglu) [inline pkg/stats/statsmgr.go:40]: I think counting file lines is impractical for huge file in new coroutine. Why not to do this when reading these data files?

**Comment by xjlgod** [comment]
Because Studio needs to know in advance how many lines the file needs to be imported , before importing.

File: pkg/csv/reader.go [hunk 11]
```diff
@@ -45,16 +45,48 @@ func (r *CSVReader) ReadLine() (base.Data, error) {
...
+	charNum := len(record)
+	bytesNum := charNum
+	for _, s := range record {
+		bytesNum += len(s)
+	}
+	return bytesNum
+}
+
+func CountFileBytes(path string, withHeader bool) (int64, error) {
+	file, err := os.Open(path)
```
Comment (veezhang) [inline pkg/csv/reader.go:74]: The statistical methods of the two functions are different.

- `countLineBytes` is the sum of bytes of the csv field.
- `CountFileBytes` is to get the file size and then subtract the header bytes.

Comment (veezhang) [reply pkg/csv/reader.go:74]:
```golang


var _ io.Reader = (*recordReader)(nil)

type recordReader struct {
	io.Reader
	lastReadBytes int
}

func (r *recordReader) Read(p []byte) (n int, err error) {
	n, err = r.Reader.Read(p)
	r.lastReadBytes = n
	return
}
func (r *CSVReader) InitReader(file *os.File) {
	r.rr = &recordReader{Reader: bufio.NewReader(file)}
	r.reader = csv.NewReader(r.rr)
	...
}

func (r *CSVReader) ReadLine() (base.Data, error) {
    line, err := r.reader.Read()
    // r.rr.lastReadBytes // get the last read bytes 
    ...
}
```

File: pkg/base/stats.go [hunk 12]
```diff
@@ -11,26 +11,29 @@ const (
...
-	Type      StatType
-	Latency   int64
-	ReqTime   int64
-	BatchSize int
-	Filename  string
+	Type          StatType
+	Latency       int64
+	ReqTime       int64
+	BatchSize     int
+	ImportedBytesNum int64
```
Comment (veezhang) [inline pkg/base/stats.go:18]: How about use `ImportedBytes` ?

File: pkg/base/types.go [hunk 13]
```diff
@@ -34,26 +34,30 @@ func (op OpType) String() string {
 type Data struct {
 	Type   OpType
 	Record Record
+	BytesNum int
```
Comment (veezhang) [inline pkg/base/types.go:37]: How about use `Bytes` ?

**Review by veezhang — APPROVED** [review]
LGTM

File: pkg/stats/statsmgr.go [hunk 14]
```diff
@@ -39,28 +63,40 @@ func (s *StatsMgr) Close() {
 }
 
 func (s *StatsMgr) updateStat(stat base.Stats) {
-	s.totalBatches++
-	s.totalCount += int64(stat.BatchSize)
-	s.totalReqTime += stat.ReqTime
-	s.totalLatency += stat.Latency
+	s.statsRW.Lock()
```
Comment (yixinglu) [inline pkg/stats/statsmgr.go:66]: Why do we need this lock?  Are these stats not be sent by one channel?

**Comment by yixinglu** [comment]
please sign the CLA firstly!

**Comment by xjlgod** [comment]
> please sign the CLA firstly!

Ok, I hava signed it.

**Review by yixinglu — APPROVED** [review]

**Comment by veezhang** [comment]
It's need not to use `reflect` as follows:

```go
package main

import (
	"bufio"
	"encoding/csv"
	"fmt"
	"io"
	"os"
)

type ProgressReader struct {
	r          io.Reader
	totalBytes int64
	bytes      int64
}

func NewProgressReader(totalBytes int64, r io.Reader) *ProgressReader {
	if totalBytes <= 0 {
		panic("the totalBytes must greater than 0")
	}
	return &ProgressReader{
		r:          r,
		totalBytes: totalBytes,
	}
}

func (r *ProgressReader) Read(p []byte) (n int, err error) {
	n, err = r.r.Read(p)
	r.bytes += int64(n)
	return n, err
}

func (r *ProgressReader) Percentage() float64 {
	return 100 * float64(r.bytes) / float64(r.totalBytes)
}

func (r *ProgressReader) PercentageOffset(n int64) float64 {
	return 100 * float64(r.bytes+n) / float64(r.totalBytes)
}

func (r *ProgressReader) Bytes() int64 {
	return r.bytes
}

func main() {
	f, err := os.Open("xxx.csv")
	if err != nil {
		panic(err)
	}
	stat, err := f.Stat()
	if err != nil {
		panic(err)
	}

	totalBytes := stat.Size()
	pr := NewProgressReader(totalBytes, f)
	br := bufio.NewReader(pr)
	r := csv.NewReader(br)

	for {
		_, err := r.Read()
		if err != nil {
			if err == io.EOF {
				break
			}
			panic(err)
		}
		fmt.Printf(
			"%.3f %.3f %d\n",
			pr.Percentage(),
			pr.PercentageOffset(int64(-br.Buffered())),
			-br.Buffered(),
		)
	}
}
```

File: pkg/csv/reader.go [hunk 15]
```diff
@@ -4,28 +4,53 @@ import (
 	"bufio"
 	"encoding/csv"
 	"fmt"
-	"os"
-
 	"github.com/vesoft-inc/nebula-importer/pkg/base"
 	"github.com/vesoft-inc/nebula-importer/pkg/config"
 	"github.com/vesoft-inc/nebula-importer/pkg/logger"
+	"io"
+	"os"
```
Comment (veezhang) [inline pkg/csv/reader.go:11]: sort it ?

File: pkg/csv/reader.go [hunk 16]
```diff
@@ -36,25 +61,37 @@ func (r *CSVReader) ReadLine() (base.Data, error) {
...
-		return base.InsertData(line), nil
+		return base.InsertData(line, n), nil
+	}
+}
+
+func (r *CSVReader) TotalBytes() (int64) {
+	for {
+		if r.initComplete {
+			return r.totalBytes
+		}
```
Comment (veezhang) [inline pkg/csv/reader.go:95]: It's inappropriate?

File: pkg/cmd/runner.go [hunk 17]
```diff
@@ -79,14 +84,33 @@ func (r *Runner) Run(yaml *config.YAMLConfig) {
...
+			fmt.Errorf("Total %d lines fail to insert into nebula graph database", statsMgr.Stats.NumFailed)))
+	}
+}
+
+func (r *Runner) QueryStats() *stats.Stats {
+	if r.stataMgr != nil {
+		if r.Readers == nil {
+			return &r.stataMgr.Stats
+		}
+		r.stataMgr.StatsCh <- base.NewOutputStats()
```
Comment (veezhang) [inline pkg/cmd/runner.go:105]: Why add chan when it can be read directly?

File: pkg/cmd/runner.go [hunk 18]
```diff
@@ -79,14 +84,33 @@ func (r *Runner) Run(yaml *config.YAMLConfig) {
...
+	if statsMgr.Stats.NumFailed > 0 {
 		r.errs = append(r.errs, errors.Wrap(errors.NotCompleteError,
-			fmt.Errorf("Total %d lines fail to insert into nebula graph database", statsMgr.NumFailed)))
+			fmt.Errorf("Total %d lines fail to insert into nebula graph database", statsMgr.Stats.NumFailed)))
+	}
+}
+
+func (r *Runner) QueryStats() *stats.Stats {
+	if r.stataMgr != nil {
+		if r.Readers == nil {
```
Comment (veezhang) [inline pkg/cmd/runner.go:102]: Why need to check the `r.Readers`?

File: pkg/cmd/runner.go [hunk 19]
```diff
@@ -79,14 +84,33 @@ func (r *Runner) Run(yaml *config.YAMLConfig) {
...
 
-	if statsMgr.NumFailed > 0 {
+	if statsMgr.Stats.NumFailed > 0 {
 		r.errs = append(r.errs, errors.Wrap(errors.NotCompleteError,
-			fmt.Errorf("Total %d lines fail to insert into nebula graph database", statsMgr.NumFailed)))
+			fmt.Errorf("Total %d lines fail to insert into nebula graph database", statsMgr.Stats.NumFailed)))
+	}
+}
+
+func (r *Runner) QueryStats() *stats.Stats {
```
Comment (veezhang) [inline pkg/cmd/runner.go:100]: How about return `stats.Stats` nor `*stats.Stats`?

**Review by veezhang — APPROVED** [review]
LGTM

**Review by yixinglu — APPROVED** [review]

**Pull request merged**